### PR TITLE
use_sim_time refactoring

### PR DESCRIPTION
**Pull request by tonynajjar**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3119) |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Gazebo |

---

## Description of contribution in a few bullet points

* I used SetParameter to set the use_sim_time parameter and removed the parameter from launch files

## Description of documentation updates required from your changes

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_bringup/launch/localization_launch.py
Comment (tonynajjar) [inline nav2_bringup/launch/localization_launch.py:104]: I first expected to use `SetUseSimTime(True)` but that didn't work. Upon further inspection of the [Action](https://github.com/ros2/launch_ros/blob/master/launch_ros/launch_ros/actions/set_use_sim_time.py#L28), I don't think it's meant to set the parameter for all the nodes in the scope. I'm not really sure what it's for.

File: nav2_bringup/launch/localization_launch.py
Comment (tonynajjar) [inline nav2_bringup/launch/localization_launch.py:134]: I would think that using the `condition` of LoadComposableNodes or GroupAction is the same

File: nav2_bringup/launch/slam_launch.py
Comment (tonynajjar) [inline nav2_bringup/launch/slam_launch.py:48]: I think we still need the RewrittenYaml here for setting the namespace?

File: nav2_bringup/launch/localization_launch.py
Comment (SteveMacenski) [inline nav2_bringup/launch/localization_launch.py:104]: does SetParameter work globally or can it be used for a single node? Also, does that work if the parameter in question is inside of the yaml file also (clearly it works if its not inside the yaml file)? I ask because this could be a really nice solution for the `yaml_filename` thing too if we can set it for only the map server if there is a non-trivial value set in the launch configuration.

File: nav2_bringup/launch/slam_launch.py
Comment (SteveMacenski) [reply nav2_bringup/launch/slam_launch.py:48]: yes

File: nav2_controller/src/controller_server.cpp
Comment (SteveMacenski) [inline nav2_controller/src/controller_server.cpp:66]: ?

File: nav2_planner/src/planner_server.cpp
Comment (SteveMacenski) [inline nav2_planner/src/planner_server.cpp:65]: ?

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp:87]: What if its not set in the yaml + they used different launch files without the set parameter?

File: nav2_bringup/launch/slam_launch.py
Comment (SteveMacenski) [inline nav2_bringup/launch/slam_launch.py:78]: Set new line for the `Node` start

Comment (SteveMacenski) [reply nav2_bringup/launch/slam_launch.py:78]: The formatting here is a little funky for the `start_map_server` action

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Generally LGTM! This is worth an update to the migration guide for - mentioning you can remove the `use_sim_time` from your param files.

File: nav2_bringup/launch/localization_launch.py
Comment (SteveMacenski) [reply nav2_bringup/launch/localization_launch.py:104]: Why not use this for each node then?

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp
Comment (tonynajjar) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp:87]: image

in case you're wondering, doesn't apply for any other param

File: nav2_bringup/launch/localization_launch.py
Comment (tonynajjar) [reply nav2_bringup/launch/localization_launch.py:104]: > does SetParameter work globally or can it be used for a single node?

It applies the parameter for all nodes within the GroupAction. See [here](https://github.com/ros2/launch_ros/blob/5434428a30a981aeee1dbadfd5b43622cef12d67/launch_ros/launch_ros/actions/set_parameter.py#L37)

> Also, does that work if the parameter in question is inside of the yaml file also

yes, the parameter in the yaml would override

For the refactoring of `yaml_filename`, I'd prefer if we can open a separate issue. I might take it but can't commit right now

File: nav2_bringup/launch/slam_launch.py
Comment (tonynajjar) [reply nav2_bringup/launch/slam_launch.py]: Yeah I know it's funky but it passes the tests, I still don't understand why we don't have an automatic python formatter in ament_lint, like we have ament_uncrustify. I found this [discussion](https://discourse.ros.org/t/python-black-formatting/13277/27?page=2) about it.

Do you use any automatic python formatter?

File: nav2_bringup/launch/localization_launch.py
Comment (SteveMacenski) [reply nav2_bringup/launch/localization_launch.py:104]: Totally understood, just asking the question, it sounds like this would be a solution for it too. Duly noted.

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp
Comment (SteveMacenski) [reply nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server_impl.hpp:87]: Huh. OK, good to know

File: nav2_bringup/launch/slam_launch.py
Comment (tonynajjar) [reply nav2_bringup/launch/slam_launch.py]: Do you usually fix those manually? Does ROS not have an auto-formatter in ament_lint?

File: nav2_controller/src/controller_server.cpp
Comment (tonynajjar) [reply nav2_controller/src/controller_server.cpp:66]: Ah yes forgot to comment here. This is similar to what is implemented in` bt_action_server_impl.hpp`. These are nodes created internally, as opposed to in the launch file, so the SetParameter doesn't seem to have any power here.

Unfortunately I did not find a way to set the `use_sim_time` parameter during the creation of the Node here because the parent node (of the controller server) is not passed `Costmap2DROS`. The next best thing I found was to set it after its creation.

I thought it wouldn't work, I assumed the Node would need to know this parameter at creation to know which kind of clock to create/subscribe to my tests proved otherwise :shrug:

File: nav2_planner/src/planner_server.cpp
Comment (tonynajjar) [reply nav2_planner/src/planner_server.cpp:65]: same as above

**Review by tonynajjar — COMMENTED** [review]
In general I think this approach to using sim time should be made more public/popular in the ROS2 community. I can't find the discussions anymore but I remember many complaints from people switching from ROS1 on having to set this parameter manually for every node.

File: nav2_controller/src/controller_server.cpp
Comment (SteveMacenski) [reply nav2_controller/src/controller_server.cpp:66]: Could you add it as a constructor argument for `Costmap2DROS` and do it in there?

File: nav2_bringup/launch/localization_launch.py
Comment (SteveMacenski) [reply nav2_bringup/launch/localization_launch.py:104]: I suppose just using the macro as Launch provides is nice, but if there's a strong reason to set it manually this way, that's OK too.

**Comment by SteveMacenski** [comment]
> In general I think this approach to using sim time should be made more public/popular in the ROS2 community. I can't find the discussions anymore but I remember many complaints from people switching from ROS1 on having to set this parameter manually for every node.


Once this is merged, you should post about it with a link!

File: nav2_controller/src/controller_server.cpp
Comment (tonynajjar) [reply nav2_controller/src/controller_server.cpp:66]: Feasible I guess but can you clarify why you like this option better? It feels a bit intrusive to me

Comment (tonynajjar) [reply nav2_controller/src/controller_server.cpp:66]: Also, to make sure I understand: you mean to pass only the use_sim_time parameter right? not the whole parent node? Should it default to False?

**Comment by SteveMacenski** [comment]
Other commits are showing up in this PR now - please rebase on main with only these comments / conflict resolutions

File: nav2_controller/src/controller_server.cpp
Comment (SteveMacenski) [reply nav2_controller/src/controller_server.cpp]: Yeah, just add the constructor argument `bool use_sim_time` which can just be passed via `get_parameter("use_sim_time").as_bool()`. Then within the costmap we can set it when the internal node is constructed as you did in the BT action. 

Setting a parameter outside like this seems a little hacky and I think just like the name / namespace of the costmap, which clock to use seems like a pretty fundamental input. Plus, something might initialize using a wrong clock then if changes are made later if `use_sim_time` is set after construction so its protecting us from a really subtle timing bug. It seems like a bad idea to have the wrong time source initialized then being changed in the application after construction. For example, other users of Costmap2DROS outside of the application servers won't know about this change and could cause weeks of headaches trying to track down why the wrong clock is used now that the example param files don't contain that argument

Less importantly, but other yaml files in the system tests are going to have the `use_sim_time` expectation to be set that this didn't update in launch, so if that's gone / deprecated, we need to update them to have it set in code - but its easier to find if you literally change the signature to find where it needs to be done.

**Comment by tonynajjar** [comment]
Not sure I see what you mean, I merged master into my branch to fix the conflicts. The changes in "Files changed" are only my changes.
Should I have rebased instead and why?

**Comment by SteveMacenski** [comment]
Oh sorry, I was a little quick on the draw there, this is fine

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp
Comment (tonynajjar) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp:92]: meaning that I need to change all creations Costmap2DROS to pass an additional use_sim_time parameter?

Comment (SteveMacenski) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp:93]: Yes, but to be fair that's only like 3 places and you already did 2 of them - only missing `costmap_2d_node.cpp`. 

Also the `name` constructor on L80 above needs the time too. That's used in more places, but setting a default there would be OK since that's mostly there for generating tests. 

> Wouldn't that be a breaking change for anyone using that class standalone (I don't know how likely that is)?

Yes, but that is what `main` is for, we need to be able to make breaking changes. The fact that we set the `use_sim_time` in launch rather than in the yaml files was already a breaking change enough that this couldn't be backported.

File: nav2_bringup/launch/localization_launch.py
Comment (SteveMacenski) [reply nav2_bringup/launch/localization_launch.py:109]: Leaving this un-resolved as a reminder to myself

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp
Comment (tonynajjar) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp:93]: Yeah I was talking more about the name-only constructor. Thanks for clarifying that this one can take a default.
Great then, I will try to wrap this up tomorrow

Comment (tonynajjar) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp:93]: > only missing costmap_2d_node.cpp

are you sure? I see the name-constructor being used there

**Comment by tonynajjar** [comment]
>  This is worth an update to the migration guide

In which migration guide should that be? Do you plan to backport this to Humble? If not, will you create the next migration guide for Iron?

**Comment by tonynajjar** [comment]
> This is worth an update to the migration guide fo

https://github.com/ros-planning/navigation.ros.org/pull/339

File: nav2_controller/src/controller_server.cpp
Comment (SteveMacenski) [inline nav2_controller/src/controller_server.cpp:65]: Last nit-pick and I'll merge this: Put the `get_parameter(` on the new line, its weird to have the line extension break there. I suspect that was a linter auto-fix. You can place `get_parameter("use_sim_time").as_bool());` all on the second line. As well for planner server, this is just a readability thing.

**Review by SteveMacenski — APPROVED** [review]

File: nav2_controller/src/controller_server.cpp
Comment (tonynajjar) [reply nav2_controller/src/controller_server.cpp]: Done. Yeah it was just automatic uncrustify. I'm pretty sure this can be avoided with some custom uncrustify config

**Review by SteveMacenski — APPROVED** [review]

**Pull request merged**
